### PR TITLE
Extract i18n bundles and enable component test cases in react

### LINKED ISSUE #7787: React : Component unit test cases are excluded from Jest configuration
##### **Overview of the issue**

Component unit test cases are excluded from Jest configuration. This can be checked by looking under `jest.conf.js` file.

##### **Motivation for or Use Case**

Generated or newly added component unit test cases are not executed

##### **Suggest a Fix**
`jest.conf.js` -->
Current:
` testMatch: ['<rootDir>/src/test/javascript/spec/**/+(*.)+(spec.ts)'],`
Required:
`  testMatch: ['<rootDir>/src/test/javascript/spec/**/+(*.)+(spec.ts?(x))'],`

We also need to add enzyme configurations for enzyme functions (shallow, mount etc) to work:
`jest.conf.js` -->
`setupFiles: ['<rootDir>/src/test/javascript/enzyme-setup.ts']`

`enzyme-setup.ts` -->
```
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
configure({ adapter: new Adapter() });
```
Above configurations enable execution of component test cases, however, few test cases fail due to usage of `require.context` in `translation.ts` as:
`en: mergeTranslations(require.context('../../i18n/en', false, /.json$/))`

As require.context is specific to webpack and Jest [discourage](https://github.com/facebook/create-react-app/issues/517) it's use, we should identify an appropriate way to resolve this issue to enable execution of component unit test cases.

##### **JHipster configuration**

Using JHipster version installed globally
Executing jhipster:info
Options: 
##### **JHipster Version(s)**

```
rgtw-4@0.0.0 /home/vishal/vishal/test/rgtw4
`-- (empty)

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.0-beta.1",
    "applicationType": "monolith",
    "baseName": "rgtw4",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "no",
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "react",
    "useSass": true,
    "clientPackageManager": "yarn",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ],
    "enableHibernateCache": false
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity BankAccount (BankAccount) {
  name String required,
  balance BigDecimal required
}
entity Label (Label) {
  label String required minlength(3)
}
entity Operation (Operation) {
  date DateTime required,
  description String,
  amount BigDecimal required
}

relationship OneToMany {
  BankAccount{operation} to Operation{bankAccount(name)}
}
relationship ManyToOne {
  BankAccount{user(login)} to User
}
relationship ManyToMany {
  Operation{label(label)} to Label{operation}
}

paginate Operation with infinite-scroll

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

git version 2.17.1

node: v8.9.1

npm: 5.5.1

yarn: 1.7.0

Docker version 18.03.1-ce, build 9ee9f40

Congratulations, JHipster execution is complete!

##### **Browsers and Operating System**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by vishal423**
- Closes #7787
- Extract and merge i18n resource bundles during webpack build.
- Load all resource bundles using axios get call. This can be further improved to lazy load bundles, however, that can be done separately.
- Enable component test cases.

Pending:
- Few configuration files, settings can be conditional based on i18n support.

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by vishal423** [comment]
@deepu105 , should I push changes for test cases fix here itself or you want separate merge request as that's dependent on this.

**Comment by deepu105** [comment]
@vishal423 since they are related push it here, so that I can test them together

File: generators/client/files-react.js [hunk 1]
```diff
@@ -404,7 +406,7 @@ const files = {
             condition: generator => generator.authenticationType !== 'oauth2',
             path: TEST_SRC_DIR,
             templates: [
-                'spec/app/modules/account/register/register.spec.tsx',
+                // 'spec/app/modules/account/register/register.spec.tsx',
```
Comment (deepu105) [inline generators/client/files-react.js]: why is this commented out?

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs [hunk 2]
```diff
@@ -51,20 +43,19 @@ const languages: any = {
...
     const previousLocale = currentLocale;
     currentLocale = store.getState().locale.currentLocale;
     if (previousLocale !== currentLocale) {
-      Storage.local.set('locale', currentLocale);
+      Storage.session.set('locale', currentLocale);
       TranslatorContext.setLocale(currentLocale);
     }
   });
   store.dispatch(setLocale(savedLocale));
-  return savedLocale;
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs]: wasnt the return used anywhere?

File: generators/client/templates/react/src/test/javascript/jest.conf.js.ejs [hunk 3]
```diff
@@ -3,10 +3,11 @@ module.exports = {
     '^.+\\.tsx?$': 'ts-jest'
   },
   rootDir: '../../../',
-  testMatch: ['<rootDir>/src/test/javascript/spec/**/+(*.)+(spec.ts)'],
+  testMatch: ['<rootDir>/src/test/javascript/spec/**/+(*.)+(spec.ts?(x))'],
   moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
   moduleNameMapper: {
-    'app/(.*)': '<rootDir>/src/main/webapp/app/$1'
+    'app/(.*)': '<rootDir>/src/main/webapp/app/$1',
+    '\\.(css|scss)$': 'identity-obj-proxy'
```
Comment (deepu105) [inline generators/client/templates/react/src/test/javascript/jest.conf.js.ejs]: why is this needed?

File: generators/client/templates/react/src/test/javascript/spec/app/shared/auth/private-route.spec.tsx.ejs [hunk 4]
```diff
@@ -26,7 +26,7 @@ describe('private-route component', () => {
     expect(error.length).toEqual(1);
     expect(error.find('.alert-danger').html()).toEqual(
       <%_ if (enableTranslation) { _%>
-      '<div class="alert alert-danger"><span>translation-not-found[error.http.403]</span></div>'
+      '<div class="alert alert-danger"><span>You are not authorized to access this page.</span></div>'
```
Comment (deepu105) [inline generators/client/templates/react/src/test/javascript/spec/app/shared/auth/private-route.spec.tsx.ejs]: nice, this was always bothering me

File: generators/client/files-react.js (hunk 1 above)
Comment (vishal423) [reply generators/client/files-react.js]: if you look into source code, then, all test cases are commented in this file. During execution Jest marks this as failure with no executable test cases. I tried to convert one test case to valid, but, later during app generation it leads to some conflicts with prettier. We can create separate issue to enable test case of this class and while doing so can uncomment this

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs (hunk 2 above)
Comment (vishal423) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs]: not used

File: generators/client/templates/react/src/test/javascript/jest.conf.js.ejs (hunk 3 above)
Comment (vishal423) [reply generators/client/templates/react/src/test/javascript/jest.conf.js.ejs]: If we don't proxy css modules, then, I noted issue in few test cases like for header.tsx

File: generators/client/templates/react/src/test/javascript/spec/app/shared/auth/private-route.spec.tsx.ejs (hunk 4 above)
Comment (vishal423) [reply generators/client/templates/react/src/test/javascript/spec/app/shared/auth/private-route.spec.tsx.ejs]: :)

**Comment by vishal423** [comment]
I may not get enough time in next few days to cover pending items. If you agree, then those can be taken care in separate issue.

File: generators/client/files-react.js (hunk 1 above)
Comment (deepu105) [reply generators/client/files-react.js]: oh, ok now I remember, I think there was some issue in this specific spec, so ya its fine

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs (hunk 2 above)
Comment (deepu105) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts.ejs]: ok

File: generators/client/templates/react/src/test/javascript/jest.conf.js.ejs (hunk 3 above)
Comment (deepu105) [reply generators/client/templates/react/src/test/javascript/jest.conf.js.ejs]: ok
